### PR TITLE
EVG-7111 enable CORS for gql requests

### PR DESCRIPTION
**Pull request by tgrander**
Enabling CORS for GQL requests requires setting the header from the GQL request in the response headers.

### CODE REVIEW

File: service/middleware.go [hunk 1]
```diff
@@ -174,12 +174,19 @@ func (uis *UIServer) isSuperUser(u gimlet.User) bool {
 
 func (uis *UIServer) setCORSHeaders(next http.HandlerFunc) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
-		if r.Method == http.MethodGet && len(uis.Settings.Ui.CORSOrigins) > 0 {
```
Comment (tgrander) [inline service/middleware.go]: only GET requests were allowed for CORS for some reason

File: service/middleware.go [hunk 2]
```diff
@@ -174,12 +174,19 @@ func (uis *UIServer) isSuperUser(u gimlet.User) bool {
...
 			requester := r.Header.Get("Origin")
+
+			// Requests from a GQL client include this header, which must be added to the response to enable CORS
+			gqlHeader := r.Header.Get("Access-Control-Request-Headers")
+
 			if util.StringSliceContains(uis.Settings.Ui.CORSOrigins, requester) {
 				w.Header().Add("Access-Control-Allow-Origin", requester)
 				w.Header().Add("Access-Control-Allow-Credentials", "true")
 				w.Header().Add("Access-Control-Allow-Headers", fmt.Sprintf("%s, %s", evergreen.APIKeyHeader, evergreen.APIUserHeader))
+				if gqlHeader != "" {
```
Comment (john-m-liu) [inline service/middleware.go]: This will overwrite the headers in the line above, these all should be added at once

File: service/ui.go [hunk 3]
```diff
@@ -279,8 +279,8 @@ func (uis *UIServer) GetServiceApp() *gimlet.APIApp {
 	}
 
 	// GraphQL
-	app.AddRoute("/graphql").Handler(playground.Handler("GraphQL playground", "/graphql/query")).Get().Wrap(needsLogin)
-	app.AddRoute("/graphql/query").Handler(handler.NewDefaultServer(graphql.NewExecutableSchema(graphql.New())).ServeHTTP).Options().Post().Get().Wrap(needsLogin)
+	app.AddRoute("/graphql").Wrap(needsLogin, allowsCORS).Handler(playground.Handler("GraphQL playground", "/graphql/query")).Get()
```
Comment (john-m-liu) [inline service/ui.go]: The needsLogin will still cause redirect problems when working with this locally. How are you planning on getting around that?

File: service/middleware.go [hunk 4]
```diff
@@ -174,12 +174,19 @@ func (uis *UIServer) isSuperUser(u gimlet.User) bool {
...
+					w.Header().Add("Access-Control-Allow-Headers", gqlHeader)
+				}
 			}
 		}
 		grip.ErrorWhen(r.Method != http.MethodGet, message.Fields{
```
Comment (john-m-liu) [inline service/middleware.go]: This log message should be removed (I think this is why the check for GET was there originally)

File: service/ui.go (hunk 3 above)
Comment (tgrander) [reply service/ui.go]: For now `needsLogin` will just need to be removed when working with gql server locally. It's not ideal but will work in the short term. There is a ticket for authentication with gql server and the details will be flushed out then.

File: service/middleware.go (hunk 2 above)
Comment (tgrander) [reply service/middleware.go]: nice! changed it

File: service/middleware.go (hunk 4 above)
Comment (tgrander) [reply service/middleware.go]: done

**Pull request merged**